File: src/Especialista.ts
// Classe abstrata que define a estrutura de um especialista

import { DadosSintomas } from "./DadosSintomas";

export abstract class Especialista {
  protected proximo: Especialista | null = null;

  // Define o próximo especialista na cadeia
  public definirProximo(proximo: Especialista): Especialista {
    this.proximo = proximo;
    return proximo;
  }

  // Método principal que tenta realizar o diagnóstico
  public analisar(dados: DadosSintomas): string {
    if (this.podeDiagnosticar(dados)) {
      return this.diagnosticar(dados);
    } else if (this.proximo) {
      return this.proximo.analisar(dados);
    } else {
      return "Nenhum diagnóstico foi possível.";
    }
  }

  // Métodos que as subclasses devem implementar
  protected abstract podeDiagnosticar(dados: DadosSintomas): boolean;
  protected abstract diagnosticar(dados: DadosSintomas): string;
}

File: SistemaDiagnostico/Patterns/src/Especialistas/Especialista.java
package Especialistas;

import Diagnostico.DadosSintomas;

import java.util.ArrayList;
import java.util.List;

public abstract class Especialista {
    protected Especialista proximo;

    public Especialista definirProximo(Especialista proximo) {
        this.proximo = proximo;
        return proximo;
    }

    // Agora retorna lista de diagnósticos possíveis
    public List<String> analisar(DadosSintomas dados) {
        List<String> resultados = new ArrayList<>();

        if (podeDiagnosticar(dados)) {
            resultados.add(diagnosticar(dados));
        }

        if (proximo != null) {
            resultados.addAll(proximo.analisar(dados));
        }

        return resultados;
    }

    protected abstract boolean podeDiagnosticar(DadosSintomas dados);
    protected abstract String diagnosticar(DadosSintomas dados);
}
